util/eventbus/bus.go

```diff
@@ -0,0 +1,216 @@
+	"reflect"
+	"slices"
+	"sync"
+
+	"tailscale.com/util/set"
+)
+
+// Bus is an event bus that distributes published events to interested
+// subscribers.
+type Bus struct {
+	write    chan any
+	stop     *goroutineShutdown
+	snapshot chan chan []any
+
+	topicsMu sync.Mutex // guards everything below.
+	topics   map[reflect.Type][]*Queue
+
+	// Used for introspection/debugging only, not in the normal event
+	// publishing path.
+	publishers set.Set[publisher]
+	queues     set.Set[*Queue]
+}
+
+// New returns a new bus. Use [PublisherOf] to make event publishers,
+// and [Bus.Queue] and [Subscribe] to make event subscribers.
+func New() *Bus {
+	ret := &Bus{
+		write:      make(chan any),
+		stop:       newGoroutineShutdown(),
+		snapshot:   make(chan chan []any),
+		topics:     map[reflect.Type][]*Queue{},
+		publishers: set.Set[publisher]{},
+		queues:     set.Set[*Queue]{},
+	}
+	go ret.pump()
+	return ret
+}
+
+func (b *Bus) pump() {
+	defer b.stop.ShutdownComplete()
+	var vals queue
+	acceptCh := func() chan any {
+		if vals.Full() {
+			return nil
+		}
+		return b.write
+	}
+	for {
+		// Drain all pending events.
+		for !vals.Empty() {
+			val := vals.Peek()
+			dests := b.dest(reflect.ValueOf(val).Type())
+			for _, d := range dests {
+			deliverOne:
+				for {
+					select {
+					case d.write <- val:
+						break deliverOne
+					case <-d.stop.Closed():
+						// Queue closed, don't block but continue
+						// delivering to others.
+						break deliverOne
+					case in := <-acceptCh():
+						vals.Add(in)
+					case <-b.stop.StartShutdown():
+						return
+					case ch := <-b.snapshot:
+						ch <- vals.Snapshot()
+					}
+				}
+			}
+			vals.Drop()
+		}
+
+		// Inbound queue empty, wait for at least 1 work item.
+		for vals.Empty() {
+			select {
+			case <-b.stop.StartShutdown():
+				return
+			case val := <-acceptCh():
+				vals.Add(val)
+			case ch := <-b.snapshot:
+				ch <- nil
+			}
+		}
+
+		// We have 1 event to dispatch. Opportunistically drain any
+		// other waiting events, to unblock publishers. Don't bother
+		// selecting on close/snapshot, those can get handled on the
+		// next dispatch cycle.
+	unblockPublishers:
+		for {
+			select {
+			case val := <-acceptCh():
+				vals.Add(val)
+			default:
+				break unblockPublishers
+			}
+		}
+	}
+}
+
+func (b *Bus) dest(t reflect.Type) []*Queue {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	return b.topics[t]
+}
+
+func (b *Bus) subscribe(t reflect.Type, q *Queue) (cancel func()) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	b.topics[t] = append(b.topics[t], q)
+	return func() {
+		b.unsubscribe(t, q)
+	}
+}
+
+func (b *Bus) unsubscribe(t reflect.Type, q *Queue) {
+	b.topicsMu.Lock()
+	defer b.topicsMu.Unlock()
+	// Topic slices are accessed by pump without holding a lock, so we
+	// have to replace the entire slice when unsubscribing.
+	// Unsubscribing should be infrequent enough that this won't
+	// matter.
+	i := slices.Index(b.topics[t], q)
+	if i == -1 {
```

Comment:
I guess it does promise to be exactly -1, but I'd probably have written `i < 0`. YMMV.

Reply:
Done.